Unify reporting and handling of Olm errors

KitsuneRal opened the pull request:
This set of changes is mainly aimed at getting rid of `throw` statements across `QOlm*` classes; along with this, there's a bit of cleanup on unnecessarily verbose or duplicating things. In particular, `QOlmError` is replaced with `OlmErrorCode` introduced  (thanks to Nico from our friendly competition for pushing that upstream) in Olm 3.2.5, and `QOlmMessage::Type` directly references constants from Olm. `RandomBuffer` is introduced to encapsulate the logic that generates randomness and purges it on scope exit (actually, Olm takes care of purging; but it doesn't hurt to have our own thing for that, too). Finally, repetitive pieces of error reporting are wrapped up in macros.

There are quite a few of new FIXMEs scattered across `QOlm*` classes, mainly calling for RAII approach to memory handling for Olm structures (that's coming just in a bit) but also questioning in a couple of places code that seems unused.

KitsuneRal committed:
Bump required Olm version to 3.2.5

This is the earliest version shipping olm/error.h. Conversely, stock
libolm that comes with Ubuntu 20.04 (version 3.1.3) is no more enough
so CI completely switches to jammy now.

KitsuneRal committed:
CI: only use Valgrind with GCC

As of the current GHA jammy image, Valgrind cannot cope with some
of Clang 14 output: https://bugs.kde.org/show_bug.cgi?id=452758

This is most likely a band-aid, before the work on switching from
Valgrind to ASan.

KitsuneRal committed:
.clang-tidy: drop readability-qualified-auto

Too many suggestions to add, e.g., * in places where it may eventually
be a smart pointer.

[skip ci]

KitsuneRal committed:
Cleanup across E2EE code

Notably:
- simplified unnecessarily verbose constructs;
- formally aligned (no re-numeration was necessary)
  QOlmMessage::Type with corresponding OLM_ constants;
- dropped QOlmSession::encryptMessageType() because it's very
  sensitive to the order of calling with QOlmSession::encrypt()
  (and encrypt() itself already calls it and returns the message
  type);
- simplify the return type of pickle() calls that can only fail due to
  an internal error;
- replace const QString& with QStringView or const QByteArray&
  where appropriate;
- use '\0' where it was meant to be instead of '0'.

KitsuneRal committed:
RandomBuffer

A convenient abstraction swallowing all the type casts and, more
importantly, cleanup on destruction (previous code only cleaned up
the buffer upon a successful call to Olm API but not upon an error).

KitsuneRal committed:
Wrap error reporting into facility macros

Facility macros to report Olm errors: QOLM_INTERNAL_ERROR[_X],
QOLM_FAIL_OR_LOG[_X]

KitsuneRal committed:
Trample Sonar warnings

TobiasFella commented on lib/room.cpp:
```diff
@@ -443,12 +444,10 @@ class Room::Private {
-            return;
-        }
-        addInboundGroupSession(currentOutboundMegolmSession->sessionId(), *sessionKey, q->localUser()->id(), "SELF"_ls);
+        currentOutboundMegolmSession->sessionKey();
```
is the point of this line to cause a QOLM_INTERNAL_ERROR if something is broken?

KitsuneRal replied:
Well, I was really just converting the previous code mechanically and didn't notice ending up with a no-op. I think this should just be dropped; if something is wrong with the session key, an internal error will occur on the very next line.

KitsuneRal committed:
Remove no-op code

KitsuneRal committed:
QOlmUtility::ed25519Verify: just return bool

It's too easy to incorrectly test the previous return type.

(cherry picked from commit 5904a61c59f0eef00aef07ef998658fd791ff139)

KitsuneRal committed:
Replace QOlmError with OlmErrorCode

QOlmError represents a subset of OlmErrorCode, and the associated
fromString() function uses undocumented strings produced inside Olm;
meanwhile OlmErrorCode is documented in its own header file. Each QOlm*
class now has lastErrorCode() next to lastError() (that, from now,
returns a textual representation straight from Olm, not QOlmError enum).

Also: including olm/error.h in e2ee/e2ee.h required some rearrangement
of the code to make sure non-E2EE configuration still builds.

KitsuneRal committed:
Merge branch 'dev' into kitsune/unify-olm-errors
